Implement Serializers for X25519 and X448

mattcaswell opened the pull request:
Provide serializers for X25519 and X448 for text, pem and der. There are
no paramater serializers because there are no parameters for these
algorithms.

levitte commented on providers/implementations/serializers/serializer_ecx.c:
```diff
@@ -0,0 +1,121 @@
+        *ecx_free = ossl_prov_get_keymgmt_free(x448_keymgmt_functions);
+        *ecx_import = ossl_prov_get_keymgmt_import(x448_keymgmt_functions);
+    }
+}
```
Why didn't I think of this???

levitte commented on providers/implementations/serializers/serializer_local.h:
```diff
@@ -49,6 +55,13 @@ int ossl_prov_prepare_dh_params(const void *dh, int nid,
+int ossl_prov_ecx_priv_to_der(const void *ecxkey, unsigned char **pder);
+
 int ossl_prov_prepare_dsa_params(const void *dsa, int nid,
                                 ASN1_STRING **pstr, int *pstrtype);
```
Nit: indent one space off

mattcaswell replied:
This is existing code...but ok

levitte replied:
Oops. Sorry

slontis commented:
looks like no-ec is broken in travis

mattcaswell commented:
Fixup pushed to address the feedback and to correct the no-ec build.

mspncp commented:
Typo in commit message: paramater -> parameter

mattcaswell commented:
Squash commit pushed fixing the typo

paulidale commented on providers/implementations/serializers/serializer_ecx.c:
```diff
@@ -0,0 +1,121 @@
+        *ecx_new = ossl_prov_get_keymgmt_new(x25519_keymgmt_functions);
+        *ecx_free = ossl_prov_get_keymgmt_free(x25519_keymgmt_functions);
+        *ecx_import = ossl_prov_get_keymgmt_import(x25519_keymgmt_functions);
+    } else {
```
Ideally, a `type == ECX_KEY_TYPE_X448` check here would be good.
There is no way to return a failure, so it's kind of a moot point.

levitte replied:
Not quite mot... Not setting the pointers would become an assertion later on. This shouldn't ever happen, but if the algo name strings are somehow wrong, we will notice!

mattcaswell replied:
Fixed.

paulidale commented on providers/implementations/serializers/serializer_ecx_priv.c:
```diff
@@ -0,0 +1,268 @@
+    struct ecx_priv_ctx_st *ctx = vctx;
+    const OSSL_PARAM *p;
+
+    if ((p = OSSL_PARAM_locate_const(params, OSSL_SERIALIZER_PARAM_CIPHER))
```
Breaking this into `p = ...;` and `if (p != NULL)` would add clarity and not increase the line count.

mattcaswell replied:
Fixed

mattcaswell commented:
Update pushed addressing feedback so far.

paulidale commented on providers/implementations/serializers/libimplementations-lib-serializer_ecx.d.tmp:
```diff
@@ -0,0 +1,25 @@
+providers/implementations/serializers/libimplementations-lib-serializer_ecx.o: \
```
oops

levitte replied:
this is a good reason not to build in the source directory...  but also to fix .gitignore

mattcaswell replied:
Fixed. Also fixed .gitignore in #11122

paulidale commented on providers/implementations/serializers/serializer_ecx_priv.c:
```diff
@@ -0,0 +1,268 @@
+    if (p != NULL) {
+        const OSSL_PARAM *propsp =
+            OSSL_PARAM_locate_const(params, OSSL_SERIALIZER_PARAM_PROPERTIES);
+        const char *props = NULL;
```
This variable is assigned NULL by default six lines down.  No need for both.

mattcaswell replied:
Fxied

paulidale commented on providers/implementations/serializers/serializer_ecx_pub.c:
```diff
@@ -0,0 +1,185 @@
+    ECX_KEY_TYPE type;
+};
+
+
```
Spurious blank line.

mattcaswell replied:
Fixed

richsalz commented:
> this is a good reason not to build in the source directory.

Woah!  That's not going to become a real thing, is it?

levitte commented:
> Woah! That's not going to become a real thing, is it?

Don't panic, it's of course still possible to build in the source directory.

mattcaswell commented:
Fixup pushed address above feedback.

kroeckx commented:
The commit message indicates this is for "text, pem and der", but I see none of them being tested. (I have no idea what text means here.)

Does follow the format from some RFC like RFC8410?

levitte commented:
Test is printouts, basically. It corresponds to the ameth print methods.

kroeckx commented:
This seems to lack documentation updates. We seem to have an
EVP_PKEY_fromdata() function, but it's not documented what that
function expects as input.

A function like EVP_PKEY_param_fromdata_settable() is making
little sense to me. That looks like something that needs to be
in the documentation, rather that first writing a test program to
see what parameters I can set.

That whole manpage needs examples.

levitte commented:
Good point, @kroeckx, thanks for pointing that out.

slontis commented:
@paulidale does this need reapproval?

mattcaswell commented:
> @paulidale does this need reapproval?

I need to find some time to look at this again in light of @kroeckx's comments.

mattcaswell commented:
I've rebased this to resolve a conflict with master and pushed some significant updates. I've expanded out the testing significantly. Previously we tested things by checking we were able to print out private keys in text and pem, as well as public keys in pem. We now generate both public and private keys in text, pem and der and compare them to a "known good" file. I've done this for all the existing serializers (not just X25519 and X448). This identified a few bugs which I've also fixed.

Additionally I've expanded out the documentation for all the various key types to document the import/export types that are available for built-in algorithms.

I've not touched EVP_PKEY_fromdata() as mentioned by @kroeckx. That's not really directly touched by this PR so probably adding examples should be covered by a subsequent PR. I see @levitte has opened an issue for this (#11131).

Note that the "known good" files referred to above are self-generated. I intend to double check these against the 1.1.1 implementation (which doesn't have all this new serializer code) to confirm that they really are good - but I've not yet done that.

Hopefully this addresses the comments that were raised. Please take another look.

slontis commented:
Travis NITs

slontis commented on providers/defltprov.c:
```diff
@@ -464,6 +464,34 @@ static const OSSL_ALGORITHM deflt_serializer[] = {
 #endif
 
+#ifndef OPENSSL_NO_EC
+    { "X25519", "default=yes,format=text,type=private",
```
these need to change from default= ?

mattcaswell replied:
Fixed

kroeckx commented:
I'm fine with examples to EVP_PKEY_fromdata being added in an other PR, but I think this PR should document the format EVP_PKEY_fromdata() is expecting.

mattcaswell commented:
> I'm fine with examples to EVP_PKEY_fromdata being added in an other PR, but I think this PR should document the format EVP_PKEY_fromdata() is expecting.

In response to your comment on this previously I added a description of all the formats to provider-keymgmt.pod (both for the serializers added by this PR, as well as all the existing serlializers). The formats are the same whether you are a provider author wanting to implement serializers that conform to those formats, or whether your are an end application author wanting to use EVP_PKEY_fromdata(). Its probably not a good idea to repeat the documentation in both places. My assumption was that the PR that updates the EVP_PKEY_fromdata() page would include a reference to the provider-keymgmt.pod page.

mattcaswell commented:
I've rebased this and (hopefully) addressed the travis nits.

> My assumption was that the PR that updates the EVP_PKEY_fromdata() page would include a reference to the provider-keymgmt.pod page.

I've now added that reference as part of this PR.

kroeckx commented on doc/man7/provider-keymgmt.pod:
```diff
@@ -256,6 +256,87 @@ OP_keymgmt_export_types() should return a constant array of descriptor
+
+The public key value.
+
+=item "priv" (B<OSSL_PKEY_PARAM_PRIV_KEY>) <integer> or <octet string>
```
Does integer make any sense here?

mattcaswell replied:
Oops. No. This is a left over from an earlier iteration where I had the description of this combined with DSA/DH version which does use an integer. Will remove.

mattcaswell replied:
Fixed.

mattcaswell commented:
I pushed a fixup addressing the one item of feedback above.

The appveyor failure looks relevant. Unfortunately I do not have access to a Windows environment right at the moment, so I won't be able to investigate it until I get back from the RSA conference. :-(

levitte commented on providers/implementations/serializers/serializer_ecx_priv.c:
```diff
@@ -0,0 +1,268 @@
+
+    ret = ossl_prov_write_priv_der_from_obj(out, ecxkey,
+                                            ctx->type == ECX_KEY_TYPE_X25519
+                                            ? EVP_PKEY_X25519 : EVP_PKEY_X448,
```
It might be worth having a local variable to get that value

mattcaswell replied:
Fixed

slontis commented:
> The appveyor failure looks relevant. Unfortunately I do not have access to a Windows environment right at the moment, so I won't be able to investigate it until I get back from the RSA conference. :-(

Looks weird..
It only happens with the first key type during X25519.pub.der
It looks like the last 10 bytes of the of bitstring are missing.. (almost looks like the length has been truncated for some reason).

mattcaswell commented:
@levitte figured it out. It's because I open the file with "r" and not "rb" and so it interprets a "1a" byte as EOF.

mattcaswell commented:
Fixups pushed to attempt to address the appveyor issue.

slontis commented on providers/implementations/serializers/serializer_common.c:
```diff
@@ -243,6 +243,36 @@ int ossl_prov_print_labeled_bignum(BIO *out, const char *label,
 }
 
+/* Number of octets per line */
+#define LABELED_BUF_PRINT_WIDTH    15
```
Why do we use such a weird number, why not 16??

mattcaswell replied:
To be consistent with the way `ASN1_buf_print` currently does this.

slontis commented:
You have a green build on one of the failing windows machines from before - so it looks like it has been addressed.

slontis commented:
Merged to master.

mattcaswell commented:
> Merged to master.

Thanks @slontis!
